Bloodsuckers, or, yknow, vampires.

Arturlang opened the pull request:
Okay, so, this PR ports Bloodsuckers, from [fulpstation](https://github.com/TheSwain/Fulpstation) by [TheSwain](https://github.com/TheSwain), which are actually way more interesting than the ones from hippie or vg or paradise. And have way less identical abilities. 
## Changelog
:cl:
add: Adds Bloodsuckers, beware.
/:cl:\
Like the species vampires, the blood that matters is inside them and slowly runs out and as they use abilities, but is increased as they progress. Passive bonuses are way more focused on, and vampires are made weak to burns, and stakes, which are added in this pr as well as the accompanying monster hunters whose sole job is to deal with monsters as such. Vampires also have to hide during a period of time where UV rays from the sun wash the station, either in a locker or a coffin, in which they have to be in to "level up", instead of having to drink blood to power up as most other vampire types. 
They can also set one coffin to make the area into their lair, which causes the coffin to be able to be locked from the inside, and allows the vampire to build vampire-styled structures, such as a candelabrum only vampires can turn on or the torturing rack which allows for conversion, which only works on mindshield implanted and antags if they say yes in the prompt that is given, who are then given minor powers and mindwashed to obey their masters orders. This PR is pretty much done and it works locally, testmerge would be nice, and as this PR is already pretty huge im gonna add additional stuff in further PRs.

Also. Not a joke PR like those pr-ed recently. Totally. Effort was made.

Arturlang committed:
Bludsuckers

Fractious reviewed (commented):
Might wanna add all the head roles to be consistant with the other antags.
`protected_jobs = list("Security Officer", "Warden", "Detective", "Head of Security", "Captain", "Head of Personnel", "Deputy")`

Ghommie commented on code/game/gamemodes/bloodsucker/bloodsucker.dm:
```diff
@@ -0,0 +1,294 @@
+		H.real_name = H.client.prefs.custom_names["human"]
+		if (H.wear_id)
+			H.wear_id.GetID().registered_name = H.real_name
+			H.wear_id.GetID().update_label()
```
It'll runtime if the owner is found without a worn ID. Replace it with this:
```
		var/obj/item/card/id/ID = H.wear_id?.GetID()
		if(ID)
			ID.registered_name = H.real_name
			ID.update_label()
```

Ghommie commented on code/game/gamemodes/bloodsucker/bloodsucker.dm:
```diff
@@ -0,0 +1,294 @@
+	if (!can_make_vassal(target,creator))
+		return FALSE
+	// Make Vassal
+	var/datum/antagonist/vassal/V = new ANTAG_DATUM_VASSAL(target.mind)
```
```suggestion
	var/datum/antagonist/vassal/V = new (target.mind)
```
will suffice

Ghommie commented on code/game/gamemodes/bloodsucker/hunter.dm:
```diff
@@ -0,0 +1,49 @@
+/datum/game_mode/proc/finalize_monster_hunters(monster_count = 4)
+	var/amEvil = TRUE // First hunter is always an evil boi
+	for(var/datum/mind/hunter in vamphunters)
+		var/datum/antagonist/vamphunter/A = new ANTAG_DATUM_HUNTER(hunter)
```
```suggestion
		var/datum/antagonist/vamphunter/A = new (hunter)
```

Ghommie commented on code/game/gamemodes/bloodsucker/bloodsucker.dm:
```diff
@@ -0,0 +1,294 @@
+
+	// [FLEDGLING]
+	if (creator)
+		A = new ANTAG_DATUM_BLOODSUCKER(bloodsucker)
```
```suggestion
		A = new (bloodsucker)
```

Ghommie commented on code/modules/antagonists/bloodsucker/bloodsucker_integration.dm:
```diff
@@ -0,0 +1,148 @@
+
+
+/mob/living/proc/HaveBloodsuckerBodyparts(var/displaymessage="") // displaymessage can be something such as "rising from death" for Torpid Sleep. givewarningto is the person receiving messages.
+	if (!getorganslot("heart"))
```
```suggestion
	if (!getorganslot(ORGAN_SLOT_HEART))
```

Ghommie commented on code/modules/antagonists/bloodsucker/bloodsucker_integration.dm:
```diff
@@ -0,0 +1,148 @@
+		if (displaymessage != "")
+			to_chat(src, "<span class='warning'>Without a heart, you are incapable of [displaymessage].</span>")
+		return FALSE
+	if (!get_bodypart("head"))
```
```suggestion
	if (!get_bodypart(BODY_ZONE_HEAD))
```

Ghommie commented on code/modules/antagonists/bloodsucker/bloodsucker_life.dm:
```diff
@@ -0,0 +1,500 @@
+
+	// Reagents (NOT Blood!)
+	if(target.reagents && target.reagents.total_volume)
+		target.reagents.reaction(owner.current, INGEST, 1 / target.reagents.total_volume) // Run Reaction: what happens when what they have mixes with what I have?
```
That's a tiny ass fraction.
Considering mobs have a reagent holder of 1000, that'd result in a 0,001 units mob reaction.
Perhaps they meant to do a 1 u reaction instead.
```suggestion
		target.reagents.reaction(owner.current, INGEST, 1) // Run Reaction: what happens when what they have mixes with what I have?
```

Arturlang replied:
I suppose thats fair

Ghommie commented on code/modules/antagonists/bloodsucker/bloodsucker_life.dm:
```diff
@@ -0,0 +1,500 @@
+	description = "<span class='nicegreen'>I have fed greedly from that which nourishes me.</span>\n"
+	mood_change = 10
+	timeout = 900
+/datum/mood_event/drankblood_bad
```
add some new line spaces between each mood event.

Ghommie commented on code/modules/antagonists/bloodsucker/bloodsucker_life.dm:
```diff
@@ -0,0 +1,500 @@
+
+/datum/antagonist/bloodsucker/proc/Torpor_Begin(amInCoffin=FALSE)
+	owner.current.stat = UNCONSCIOUS
+	owner.current.fakedeath("bloodsucker") // Come after UNCONSCIOUS or else it fails
```
You might want to convert this trait source text string into a define. Your call.

Arturlang replied:
What would it help with? Just wondering or is it just for neatness?

Ghommie replied:
Also voids typos.

Arturlang replied:
Rather not, tested it and it breaks stuff

Ghommie commented on code/modules/antagonists/bloodsucker/bloodsucker_powers.dm:
```diff
@@ -0,0 +1,413 @@
+	// Incap?
+	if(must_be_capacitated)
+		var/mob/living/L = owner
+		if (L.incapacitated(ignore_restraints=TRUE,ignore_grab=TRUE) || L.lying && !can_be_immobilized)
```
```suggestion
		if (L.incapacitated(TRUE, TRUE) || L.lying && !can_be_immobilized)
```

Ghommie commented on code/modules/antagonists/bloodsucker/bloodsucker_powers.dm:
```diff
@@ -0,0 +1,413 @@
+
+/datum/action/bloodsucker/targeted/CheckCanUse(display_error)
+	if (!..())
+		return FALSE
```
```
	. = ..()
	if(!.)
		return
```

Ghommie commented on code/modules/antagonists/bloodsucker/powers/bs_cloak.dm:
```diff
@@ -0,0 +1,68 @@
+
+/datum/action/bloodsucker/cloak/CheckCanUse(display_error)
+	if(!..(display_error))// DEFAULT CHECKS
+		return FALSE
```
```
	. = ..()
	if(!.)
		return
```

Ghommie commented on code/modules/antagonists/bloodsucker/powers/bs_feed.dm:
```diff
@@ -0,0 +1,342 @@
+
+/datum/action/bloodsucker/feed/CheckCanUse(display_error)
+	if(!..(display_error))// DEFAULT CHECKS
+		return FALSE
```
idem

Ghommie commented on code/modules/antagonists/bloodsucker/powers/bs_haste.dm:
```diff
@@ -0,0 +1,92 @@
+
+/datum/action/bloodsucker/targeted/haste/CheckCanUse(display_error)
+	if(!..(display_error))// DEFAULT CHECKS
+		return FALSE
```
same

Ghommie commented on code/modules/antagonists/bloodsucker/powers/bs_gohome.dm:
```diff
@@ -0,0 +1,140 @@
+
+/datum/action/bloodsucker/gohome/CheckCanUse(display_error)
+	if(!..(display_error))// DEFAULT CHECKS
+		return FALSE
```
same

Ghommie commented on code/modules/antagonists/bloodsucker/powers/bs_brawn.dm:
```diff
@@ -0,0 +1,187 @@
+
+/datum/action/bloodsucker/targeted/brawn/CheckCanUse(display_error)
+	if(!..(display_error))// DEFAULT CHECKS
+		return FALSE
```
yea, you know the drill, needless to mention further abilities' CheckCanUse.
Same applies to the `if(!..())`

Arturlang replied:
No difference needed? Change it like the rest?

Ghommie replied:
Yea

Ghommie commented on code/modules/antagonists/bloodsucker/powers/bs_veil.dm:
```diff
@@ -0,0 +1,165 @@
+	var/prev_facial_hair_color
+	var/prev_underwear
+	var/prev_undershirt
+	var/prev_socks
```
I suppose you'd could also store these undergarments' colors into separate vars, since that's a feature here.

Arturlang replied:
Ill do this tommorow.

Arturlang replied:
Done

Arturlang committed:
Whoops forgot about citadel restricted jobs

Ghommie commented on code/modules/antagonists/bloodsucker/powers/v_distress.dm:
```diff
@@ -0,0 +1,13 @@
+
+
+/*
+/datum/action/bloodsucker/vassal/distress
```
This is quite the anathema of all the commented stuff on this PR.
I'm not going to ask you to remove those comments, most of them make sense and it'd be too pretentious to expect newly introduced features to be lavishly complete anyway.
But please move this commented wip into another file.

Arturlang replied:
Ill just remove it, it was previously just a empty action

Ghommie reviewed (commented):
At a quick glance, there doesn't seem to be many issues for what this adds, without dwelling too much into the logic of each bloodsucking power and its peculiarities.
Hoping the best on the testmerge, gj Swain.

Arturlang committed:
Reviewer fixes and spacing

Arturlang committed:
Ditto

Arturlang committed:
Merge remote-tracking branch 'upstream/master' into Bloodsuckers

Arturlang committed:
Iconfix and co


Traits instead of procs for fakedeath

Quickfix

Whoops this too

Loadsa fixes

Nerfs stamina regen

Too slow

Hm

Arturlang committed:
Fixeesss

Arturlang committed:
Fixes bad buds and toxin healing

Ghommie commented on code/__DEFINES/status_effects.dm:
```diff
@@ -81,6 +81,9 @@
 
 #define STATUS_EFFECT_PENIS_ENLARGEMENT /datum/status_effect/chem/penis_enlarger //More applied slowdown, just like the above.
+
+#define STATUS_EFFECT_NO_COMBAT_MODE /datum/status_effect/no_combat_mode/ //Wont allow combat mode and will disable it
```
I guess you could make the electrode status effect a subtype of this then, if you wish to.
Also remove the last slash from the typepath.

Arturlang replied:
That was my plan originally, though i want to ask bhjin for permission first if theyre okay with it

Arturlang replied:
Got permission and did it

Ghommie commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -240,6 +240,35 @@ GLOBAL_LIST_INIT(wood_recipes, list ( \
+				 "<span class='notice'>You begin whittling [src] into a sharp point at one end.</span>", \
+				 "<span class='italics'>You hear wood carving.</span>")
+		// 8 Second Timer
+		if(!do_mob(user, src, 80))
```
do_mob is used when both the user and the target are mobs. I guess this should be
```suggestion
		if(!do_after(user, 80, TRUE, src))
```
Though, currently, there is no such big difference between do_mob and do_after...

Ghommie commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -240,6 +240,35 @@ GLOBAL_LIST_INIT(wood_recipes, list ( \
+		if(!do_mob(user, src, 80))
+			return
+		// Make Stake
+		var/obj/item/stake/basic/new_item = new(usr.loc)
```
```suggestion
		var/obj/item/stake/basic/new_item = new(user.loc)
```

Arturlang committed:
Review fixes and tweaks

Arturlang committed:
Merge remote-tracking branch 'upstream/master' into Bloodsuckers

Arturlang committed:
More fixes. Urgh

Arturlang committed:
Whoops

Arturlang committed:
Mindshield fix and thrall buff

Linzolle commented on code/modules/antagonists/bloodsucker/bloodsucker_integration.dm:
```diff
@@ -0,0 +1,152 @@
+
+	// Normal Creatures:
+	if(!mind || !mind.has_antag_datum(ANTAG_DATUM_BLOODSUCKER))
+		return blood_volume < BLOOD_VOLUME_SAFE
```
```suggestion
		return blood_volume < (BLOOD_VOLUME_SAFE*blood_ratio)
```
To make sure that this is taken into account.

Linzolle commented on code/modules/antagonists/bloodsucker/bloodsucker_integration.dm:
```diff
@@ -0,0 +1,152 @@
+		return FALSE
+
+	// If a Bloodsucker is malnourished, AND if his temperature matches his surroundings (aka he hasn't fed recently and looks COLD)...
+	return  blood_volume < BLOOD_VOLUME_OKAY // && !(bodytemperature <= get_temperature() + 2)
```
```suggestion
	return  blood_volume < (BLOOD_VOLUME_OKAY*blood_ratio) // && !(bodytemperature <= get_temperature() + 2)
```
again

Arturlang committed:
Mooaar fixeees

Ghommie commented on code/modules/antagonists/bloodsucker/datum_vassal.dm:
```diff
@@ -0,0 +1,152 @@
+/datum/antagonist/vassal/proc/remove_thrall_eyes()
+	// Eyes
+	var/obj/item/organ/eyes/E = new
+		E.Insert(owner.current)
```
```suggestion
	E.Insert(owner.current)
```

Ghommie commented on code/modules/antagonists/bloodsucker/items/bloodsucker_organs.dm:
```diff
@@ -0,0 +1,75 @@
+/datum/antagonist/bloodsucker/proc/RemoveVampOrgans()
+	// Heart
+	var/obj/item/organ/heart/H = new
+		H.Insert(owner.current)
```
```suggestion
	H.Insert(owner.current)
```

Ghommie commented on code/modules/antagonists/bloodsucker/items/bloodsucker_organs.dm:
```diff
@@ -0,0 +1,75 @@
+		H.Insert(owner.current)
+	// Eyes
+	var/obj/item/organ/eyes/E = new
+		E.Insert(owner.current)
```
```suggestion
	E.Insert(owner.current)
```

Ghommie commented on code/modules/antagonists/bloodsucker/items/bloodsucker_organs.dm:
```diff
@@ -0,0 +1,75 @@
+/obj/item/organ/heart/proc/HeartStrengthMessage()
+	if (beating)
+		return "a healthy"
+	return "<span class='danger'>an unstable</span>"
```
Should be moved this to its specific file.

Arturlang committed:
Reviewer and linter fixes, and veil tweak

Arturlang committed:
Hopefully this fixes the linter errors

Arturlang committed:
Slaughter demons shouldnt bleed
